processor/lsmintervalprocessor/internal/merger/value.go

```diff
@@ -0,0 +1,796 @@
+	"github.com/elastic/opentelemetry-collector-components/processor/lsmintervalprocessor/internal/merger/limits"
+)
+
+const (
+	resourceLimitsEncodingPrefix = "_resource"
+	overflowMetricName           = "_other"
+	overflowMetricDesc           = "Overflow count due to datapoints limit"
+)
+
+// Not safe for concurrent use.
+type Value struct {
+	resourceLimitCfg config.LimitConfig
+	scopeLimitCfg    config.LimitConfig
+	scopeDPLimitCfg  config.LimitConfig
+
+	source pmetric.Metrics
+	// Keeps track of resource metrics overflow
+	resourceLimits *limits.Tracker
+
+	// Lookup tables created from source
+	resLookup      map[identity.Resource]resourceMetrics
+	scopeLookup    map[identity.Scope]scopeMetrics
+	metricLookup   map[identity.Metric]metric
+	numberLookup   map[identity.Stream]numberDataPoint
+	summaryLookup  map[identity.Stream]summaryDataPoint
+	histoLookup    map[identity.Stream]histogramDataPoint
+	expHistoLookup map[identity.Stream]exponentialHistogramDataPoint
+}
+
+type resourceMetrics struct {
+	pmetric.ResourceMetrics
+
+	// Keeps track of scope overflows within each resource metric
+	scopeLimits *limits.Tracker
+}
+
+type scopeMetrics struct {
+	pmetric.ScopeMetrics
+
+	// Keeps track of datapoints limits within each scope metric
+	datapointsLimits *limits.Tracker
+}
+
+type metric = pmetric.Metric
+
+type numberDataPoint = pmetric.NumberDataPoint
+
+type summaryDataPoint = pmetric.SummaryDataPoint
+
+type histogramDataPoint = pmetric.HistogramDataPoint
+
+type exponentialHistogramDataPoint = pmetric.ExponentialHistogramDataPoint
+
+// NewValue creates a new instance of the value with the configured limiters.
+func NewValue(resLimit, scopeLimit, dpLimit config.LimitConfig) Value {
+	return Value{
+		resourceLimitCfg: resLimit,
+		scopeLimitCfg:    scopeLimit,
+		scopeDPLimitCfg:  dpLimit,
+		source:           pmetric.NewMetrics(),
+		resourceLimits:   limits.NewTracker(resLimit.MaxCardinality),
+		resLookup:        make(map[identity.Resource]resourceMetrics),
+		scopeLookup:      make(map[identity.Scope]scopeMetrics),
+		metricLookup:     make(map[identity.Metric]metric),
+		numberLookup:     make(map[identity.Stream]numberDataPoint),
+		summaryLookup:    make(map[identity.Stream]summaryDataPoint),
+		histoLookup:      make(map[identity.Stream]histogramDataPoint),
+		expHistoLookup:   make(map[identity.Stream]exponentialHistogramDataPoint),
+	}
+}
+
+// MarshalProto marshals the value into binary. Before marshaling the metric,
+// the overflow values are encoded as attributes with pre-defined keys.
+func (s *Value) MarshalProto() ([]byte, error) {
+	rms := s.source.ResourceMetrics()
+	if rms.Len() > 0 {
+		// Encode resource tracker at the 0th resource metrics
+		// TODO (lahsivjar): Is this safe? We don't ever remove
+		// resource metrics so it should be but best to check.
+		// Also, the limits checker should ensure max cardinality
+		// is greater than zero.
```

Comment:
Another approach we could take that does not involve piggybacking on attributes would be to first encode the non-overflowed ResourceMetrics, and then encode a tracker for each node in the tree in traversal order. i.e. one for the Resource, then for each Scope in that resource, etc.

That may increase the storage size slightly when there are no/few overflows, but would avoid having to iterate through the attributes when unmarshalling. If we were to do this, we would probably want to add some versioning to the format, e.g. so we can add another level of limits.

Did you consider this? Not sure if it's better at all, what you've done is fairly straightforward - I mostly worry about the performance cost of iterating through attributes when unmarshalling individual batches.

Reply:
Updated the code to encode tracker separately. This resulted in lesser allocation and better performance. The encoding is pretty straightforward right now and there are some scopes for further improvements but I don't think they would be a bottleneck.